build_binary.py

```diff
@@ -0,0 +1,519 @@
+
+
+def _darwin_archive_name(version: Version) -> str:
+    # TODO: once on 3.9+ we can target lower mac versions (weak linking)
+    macos, _, _ = platform.mac_ver()
+    macos = macos.replace('.', '_')
+    return f'python-{version.s}-macosx_{macos}_{platform.machine()}.tgz'
+
+
+class _Relink(Protocol):
+    def __call__(self, filename: str, libdir: str, *, set_name: bool) -> None:
+        ...
+
+
+class Platform(NamedTuple):
+    setup_deps: Callable[[Version], int]
+    configure_args: Callable[[], tuple[str, ...]]
+    modify_env: Callable[[MutableMapping[str, str]], None]
+    linked: Callable[[str], list[str]]
+    relink: _Relink
+    archive_name: Callable[[Version], str]
+
+
+plats = {
+    'linux': Platform(
+        setup_deps=_linux_setup_deps,
+        configure_args=_linux_configure_args,
+        modify_env=_linux_modify_env,
+        linked=_linux_linked,
+        relink=_linux_relink,
+        archive_name=_linux_archive_name,
+    ),
+    'darwin': Platform(
+        setup_deps=_darwin_setup_deps,
+        configure_args=_darwin_configure_args,
+        modify_env=_darwin_modify_env,
+        linked=_darwin_linked,
+        relink=_darwin_relink,
+        archive_name=_darwin_archive_name,
+    ),
+}
+plat = plats[sys.platform]
+
+
+def _sanitize_environ(environ: MutableMapping[str, str]) -> None:
+    # remove any environ variables that may interfere with build
+    for k in ('CFLAGS', 'CPPFLAGS', 'LDFLAGS', 'PKG_CONFIG_PATH'):
+        environ.pop(k, None)
+
+    # prevent homebrew from wasting time updating
+    environ['HOMEBREW_NO_AUTO_UPDATE'] = '1'
+    # set PATH to a minimal path (avoid homebrew / local installs)
+    environ['PATH'] = '/usr/bin:/bin:/usr/sbin:/sbin'
+
+
+def _download(py: Python, target: str) -> None:
+    req = urllib.request.urlopen(py.url)
+    with open(target, 'wb') as f:
+        shutil.copyfileobj(req, f)
+
+    checksum = hashlib.sha256()
+    with open(target, 'rb') as f:
+        bts = f.read(4096)
+        while bts:
+            checksum.update(bts)
+            bts = f.read(4096)
+
+    if not secrets.compare_digest(checksum.hexdigest(), py.sha256):
+        raise SystemExit(
+            f'checksum mismatch:\n'
+            f'- got: {checksum.hexdigest()}\n'
+            f'- expected: {py.sha256}\n',
+        )
+
+
+def _extract_strip_1(tgz: str, target: str) -> None:
+    # `tar --strip-components=1` would be faster
+    # but requires gnu tar (not reliably available on macos)
+    os.makedirs(target, exist_ok=True)
+    with tarfile.open(tgz) as tarf:
+        members = []
+        for member in tarf.getmembers():
+            _, _, member.path = member.path.partition('/')
+            members.append(member)
+        tarf.extractall(target, members=members)
+
+
+def _build(build_dir: str, prefix: str) -> int:
+    if subprocess.call(
+        (
+            './configure',
+            '--prefix', prefix,
+            '--without-ensurepip',
+            '--enable-optimizations',
```

Comment:
What compiler's being used? I assume gcc's in the builder image, and if so, can we pin it to a specific version (gcc-12?)

Reply:
for linux it's a pretty old version of gcc (6.3.0-18+deb9u1: linked to the libc version in the image) -- if we're more confident that we run on modern linuxes we can probably ditch the manylinux image and use like `ubuntu:bionic` or `ubuntu:focal` (which would change our minimum libc version)

Reply:
In production, we run Debian buster (getsentry Dockerfile is based on `python:3.8.12-slim-buster`):

```
# ldd --version
ldd (Debian GLIBC 2.28-10) 2.28
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

Reply:
cool, will update in a followup 👍